src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -686,13 +708,24 @@ void FreezeBase::freeze_fast_copy(stackChunkOop chunk, int chunk_start_sp CONT_J
 
   // patch return pc of the bottom-most frozen frame (now in the chunk) with the actual caller's return address
   intptr_t* chunk_bottom_sp = chunk_top + cont_size() - _cont.argsize() - frame::metadata_words_at_top;
-  assert(_empty || *(address*)(chunk_bottom_sp-frame::sender_sp_ret_address_offset()) == StubRoutines::cont_returnBarrier(), "");
-  *(address*)(chunk_bottom_sp - frame::sender_sp_ret_address_offset()) = chunk->pc();
+#ifdef ASSERT
+  if (!_empty) {
+    intptr_t* retaddr_slot = (chunk_bottom_sp
+                              - frame::sender_sp_ret_address_offset());
+    assert(ContinuationHelper::return_address_at(retaddr_slot)
+           == StubRoutines::cont_returnBarrier(),
+           "should be the continuation return barrier");
+  }
+#endif
+  ContinuationHelper::patch_return_address_at(
+    chunk_bottom_sp - frame::sender_sp_ret_address_offset(),
```

Comment:
How about reusing retaddr_slot here?

Reply:
Yes, we can.

In the latest commit, I introduced one specific variable name `chunk_bottom_retaddr_slot` rather than using the common `retaddr_slot`.
Because the common `retaddr_slot` is used in several sites in this file and I'd like to use it in the limited assertion scope. That's why I put the assertions at lines 600/621/1887 in extra braces.